File: ios/Jazzify/Defense/Engine/DefenseVoicingKeys.swift
import Foundation

enum DefensePlayStyle: String, Sendable, Equatable {
    case phrase
    case chordVoicing = "chord_voicing"
}

enum DefenseVoicingKeyMode: String, Sendable, Equatable {
    case order
    case random
}

struct DefenseVoicingKeyState: Equatable, Sendable {
    let mode: DefenseVoicingKeyMode
    let keys: [String]
    let index: Int

    var currentKey: String {
        keys[safe: index] ?? keys.first ?? "C"
    }
}

enum DefenseVoicingKeys {
    static func isChordVoicingStage(_ stage: DefenseStageDefinition) -> Bool {
        stage.playStyle == .chordVoicing
    }

    static func buildOrderedKeyCycle(startKey: String) -> [String] {
        let all = TrainingTwoHandVoicingTables.allMajorKeys
        guard let startIndex = all.firstIndex(of: startKey) else { return all }
        return Array(all[startIndex...]) + Array(all[..<startIndex])
    }

    static func buildRandomKeyBag(avoidKey: String?) -> [String] {
        var bag = TrainingTwoHandVoicingTables.allMajorKeys
        bag.shuffle()
        if let avoidKey,
           bag.first == avoidKey,
           bag.count > 1 {
            let swapIndex = Int.random(in: 1..<bag.count)
            bag.swapAt(0, swapIndex)
        }
        return bag
    }

    static func createInitialKeyState(mode: DefenseVoicingKeyMode, startKey: String) -> DefenseVoicingKeyState {
        let keys = mode == .order ? buildOrderedKeyCycle(startKey: startKey) : buildRandomKeyBag(avoidKey: nil)
        return DefenseVoicingKeyState(mode: mode, keys: keys, index: 0)
    }

    static func advanceKeyState(_ state: DefenseVoicingKeyState) -> DefenseVoicingKeyState {
        let nextIndex = state.index + 1
        if nextIndex < state.keys.count {
            return DefenseVoicingKeyState(mode: state.mode, keys: state.keys, index: nextIndex)
        }
        if state.mode == .order {
            return DefenseVoicingKeyState(mode: state.mode, keys: state.keys, index: 0)
        }
        return DefenseVoicingKeyState(
            mode: state.mode,
            keys: buildRandomKeyBag(avoidKey: state.currentKey),
            index: 0
        )
    }

    static func stepKeyState(_ state: DefenseVoicingKeyState, delta: Int) -> DefenseVoicingKeyState {
        guard !state.keys.isEmpty else { return state }
        let count = state.keys.count
        let nextIndex = (state.index + delta % count + count) % count
        return DefenseVoicingKeyState(mode: state.mode, keys: state.keys, index: nextIndex)
    }

    static func buildActivePhrases(stage: DefenseStageDefinition, keyState: DefenseVoicingKeyState) -> [DefensePhraseDefinition] {
        guard isChordVoicingStage(stage),
              let template = stage.phrases.first,
              let referenceKey = stage.voicingLowestKey,
              let minNote = stage.voicingMinLowestNote
        else {
            return stage.phrases
        }
        return [
            transposePhrase(
                template,
                referenceKey: referenceKey,
                targetKey: keyState.currentKey,
                minLowestNote: minNote
            ),
        ]
    }

    private static func transposePhrase(
        _ template: DefensePhraseDefinition,
        referenceKey: String,
        targetKey: String,
        minLowestNote: String
    ) -> DefensePhraseDefinition {
        let keyFifths = TrainingTwoHandVoicingTables.abaSet(key: targetKey)?.keyFifths ?? template.keyFifths
        let chords = template.chords.map {
            transposeChord($0, referenceKey: referenceKey, targetKey: targetKey, minLowestNote: minLowestNote)
        }
        return DefensePhraseDefinition(
            id: template.id,
            orderIndex: template.orderIndex,
            title: targetKey,
            audioUrl: template.audioUrl,
            loopStartMeasure: template.loopStartMeasure,
            loopEndMeasure: template.loopEndMeasure,
            keyFifths: keyFifths,
            requiredCompletionCount: template.requiredCompletionCount,
            chords: chords
        )
    }

    private static func transposeIntervalSpec(
        referenceKey: String,
        targetKey: String
    ) -> TrainingMusicTheory.IntervalSpec? {
        guard referenceKey != targetKey else { return nil }
        return TrainingMusicTheory.ascendingInterval(fromRoot: referenceKey, toRoot: targetKey)
    }

    private static func transposeChord(
        _ chord: SurvivalPhraseChord,
        referenceKey: String,
        targetKey: String,
        minLowestNote: String
    ) -> SurvivalPhraseChord {
        let interval = transposeIntervalSpec(referenceKey: referenceKey, targetKey: targetKey)
        let rawNames = chord.notes.map(\.noteName)
        let transposed = interval == nil
            ? rawNames
            : rawNames.compactMap { transposeNoteName($0, by: interval) }
        let minMidi = TrainingMusicTheory.parseVoicingMidi(minLowestNote) ?? 48
        let repositioned = placeLowestInOctaveAbove(transposed, minMidi: minMidi)
        let notes = zip(chord.notes.indices, chord.notes).map { index, note in
            let name = repositioned[safe: index] ?? note.noteName
            let midi = TrainingMusicTheory.parseVoicingMidi(name) ?? note.pitchMidi
            let staffLabel = note.staffChordName.flatMap { label -> String? in
                guard interval != nil else { return label }
                return transposeChordSymbol(label, referenceKey: referenceKey, targetKey: targetKey)
            }
            return SurvivalPhraseChordNote(
                orderIndex: note.orderIndex,
                pitchMidi: midi,
                pitchClass: ((midi % 12) + 12) % 12,
                noteName: name,
                staff: midi < 60 ? 2 : 1,
                stepIndex: note.stepIndex,
                staffChordName: staffLabel
            )
        }
        let chordName = interval == nil
            ? chord.chordName
            : transposeChordSymbol(chord.chordName, referenceKey: referenceKey, targetKey: targetKey)
        return SurvivalPhraseChord(
            id: chord.id,
            orderIndex: chord.orderIndex,
            chordName: chordName,
            measureNumber: chord.measureNumber,
            notes: notes
        )
    }

    private static func transposeNoteName(
        _ name: String,
        by interval: TrainingMusicTheory.IntervalSpec?
    ) -> String? {
        guard let interval else { return name }
        guard let note = TrainingMusicTheory.parseSpelled(name) else { return nil }
        return TrainingMusicTheory.transpose(note, by: interval).name
    }

    private static func transposeChordSymbol(_ symbol: String, referenceKey: String, targetKey: String) -> String {
        let trimmed = symbol.trimmingCharacters(in: .whitespacesAndNewlines)
        guard referenceKey != targetKey,
              let interval = transposeIntervalSpec(referenceKey: referenceKey, targetKey: targetKey)
        else {
            return symbol
        }
        let parts = trimmed.components(separatedBy: " | ")
        if parts.count > 1 {
            return parts.map { transposeSingleChordSymbol($0, by: interval) }.joined(separator: " | ")
        }
        return transposeSingleChordSymbol(trimmed, by: interval)
    }

    private static func transposeSingleChordSymbol(
        _ symbol: String,
        by interval: TrainingMusicTheory.IntervalSpec
    ) -> String {
        let trimmed = symbol.trimmingCharacters(in: .whitespacesAndNewlines)
        guard let match = trimmed.range(of: #"^[A-G](?:bb|##|b|#|x)?"#, options: .regularExpression) else {
            return symbol
        }
        let root = String(trimmed[match])
        let suffix = String(trimmed[match.upperBound...])
        guard let rootPc = TrainingMusicTheory.parsePitchClassName(root) else {
            return symbol
        }
        let spelled = TrainingMusicTheory.SpelledNote(
            letterIndex: rootPc.letterIndex,
            alter: rootPc.alter,
            octave: 4
        )
        let transposed = TrainingMusicTheory.transpose(spelled, by: interval)
        return transposed.pitchName + suffix
    }

    private static func placeLowestInOctaveAbove(_ names: [String], minMidi: Int) -> [String] {
        let notes = names.compactMap { TrainingMusicTheory.parseSpelled($0) }
        guard !notes.isEmpty else { return names }
        return TrainingMusicTheory.placeLowestInOctaveAbove(notes, minMidi: minMidi).map(\.name)
    }
}

private extension Array {
    subscript(safe index: Int) -> Element? {
        guard indices.contains(index) else { return nil }
        return self[index]
    }
}

File: src/game/defense/defenseVoicingKeys.ts
/**
 * Chord voicing mode: 12-key cycle, transposition, octave repositioning.
 */
import { distance, transpose } from 'tonal';

import type { DefensePhrase, DefensePhraseChord, DefenseStage } from '@/game/defense/defenseTypes';
import {
  ALL_MAJOR_KEYS,
  KEY_FIFTHS_BY_MAJOR,
  type MajorKey,
} from '@/utils/twoHandVoicingIntermediateCourse';
import { parseVoicingNoteName } from '@/utils/voicingMusicXml';

export type DefenseVoicingKeyMode = 'order' | 'random';

export const DEFENSE_MAJOR_KEYS: readonly MajorKey[] = ALL_MAJOR_KEYS;

export const isDefenseMajorKey = (value: string): value is MajorKey => (
  DEFENSE_MAJOR_KEYS.includes(value as MajorKey)
);

export const isDefenseChordVoicingStage = (
  stage: Pick<DefenseStage, 'playStyle'>,
): boolean => stage.playStyle === 'chord_voicing';

const normalizeSpelling = (name: string): string => name.replace(/##/g, 'x');

const midiOf = (name: string): number => parseVoicingNoteName(name).midi;

const shiftOctave = (name: string, delta: number): string => {
  const parsed = parseVoicingNoteName(name);
  const accidental = parsed.alter === 2 ? 'x'
    : parsed.alter === 1 ? '#'
      : parsed.alter === -1 ? 'b'
        : parsed.alter === -2 ? 'bb'
          : '';
  return `${parsed.step}${accidental}${parsed.octave + delta}`;
};

/** Lowest note in [minMidi, minMidi + 12) via parallel shift. */
export const placeLowestInOctaveAbove = (
  names: readonly string[],
  minMidi: number,
): string[] => {
  if (names.length === 0) return [];
  let lowest = Number.POSITIVE_INFINITY;
  for (const name of names) {
    const midi = midiOf(name);
    if (midi < lowest) lowest = midi;
  }
  const octaveDelta = Math.ceil((minMidi - lowest) / 12);
  if (octaveDelta === 0) return [...names];
  return names.map((name) => shiftOctave(name, octaveDelta));
};

export const buildOrderedKeyCycle = (startKey: MajorKey): readonly MajorKey[] => {
  const startIndex = ALL_MAJOR_KEYS.indexOf(startKey);
  if (startIndex < 0) {
    return ALL_MAJOR_KEYS;
  }
  return [
    ...ALL_MAJOR_KEYS.slice(startIndex),
    ...ALL_MAJOR_KEYS.slice(0, startIndex),
  ];
};

const shuffleInPlace = <T,>(items: T[]): void => {
  for (let index = items.length - 1; index > 0; index -= 1) {
    const swapIndex = Math.floor(Math.random() * (index + 1));
    const tmp = items[index];
    items[index] = items[swapIndex] ?? tmp;
    items[swapIndex] = tmp;
  }
};

/** Random bag of 12 keys; first key differs from avoidKey when provided. */
export const buildRandomKeyBag = (avoidKey: MajorKey | null): readonly MajorKey[] => {
  const bag = [...ALL_MAJOR_KEYS];
  shuffleInPlace(bag);
  if (avoidKey != null && bag[0] === avoidKey && bag.length > 1) {
    const swapIndex = 1 + Math.floor(Math.random() * (bag.length - 1));
    const tmp = bag[0];
    bag[0] = bag[swapIndex] ?? tmp;
    bag[swapIndex] = tmp;
  }
  return bag;
};

export interface DefenseVoicingKeyState {
  readonly mode: DefenseVoicingKeyMode;
  readonly keys: readonly MajorKey[];
  readonly index: number;
}

export const createInitialVoicingKeyState = (
  mode: DefenseVoicingKeyMode,
  startKey: MajorKey,
): DefenseVoicingKeyState => ({
  mode,
  keys: mode === 'order' ? buildOrderedKeyCycle(startKey) : buildRandomKeyBag(null),
  index: 0,
});

export const currentVoicingKey = (state: DefenseVoicingKeyState): MajorKey => (
  state.keys[state.index] ?? state.keys[0] ?? 'C'
);

export const advanceVoicingKey = (
  state: DefenseVoicingKeyState,
): DefenseVoicingKeyState => {
  const nextIndex = state.index + 1;
  if (nextIndex < state.keys.length) {
    return { ...state, index: nextIndex };
  }
  if (state.mode === 'order') {
    return { ...state, index: 0 };
  }
  const lastKey = currentVoicingKey(state);
  return {
    mode: state.mode,
    keys: buildRandomKeyBag(lastKey),
    index: 0,
  };
};

export const stepVoicingKey = (
  state: DefenseVoicingKeyState,
  delta: -1 | 1,
): DefenseVoicingKeyState => {
  const count = state.keys.length;
  if (count === 0) return state;
  const nextIndex = (state.index + delta + count) % count;
  return { ...state, index: nextIndex };
};

const transposeNoteName = (name: string, interval: string): string => {
  const transposed = transpose(name, interval);
  if (!transposed) {
    throw new Error(`transpose failed: ${name} + ${interval}`);
  }
  return normalizeSpelling(transposed);
};

const transposeSingleChordSymbol = (chordName: string, interval: string): string => {
  const match = chordName.trim().match(/^([A-G](?:bb|##|b|#|x)?)([\s\S]*)$/);
  if (!match) return chordName;
  const root = interval === '1P'
    ? match[1]
    : transpose(match[1] ?? 'C', interval);
  if (!root) return chordName;
  return `${normalizeSpelling(root)}${match[2] ?? ''}`;
};

const transposeChordSymbol = (chordName: string, interval: string): string => {
  if (interval === '1P') return chordName;
  const parts = chordName.split(' | ');
  if (parts.length > 1) {
    return parts.map((part) => transposeSingleChordSymbol(part, interval)).join(' | ');
  }
  return transposeSingleChordSymbol(chordName, interval);
};

const staffForMidi = (midi: number): 1 | 2 => (midi < 60 ? 2 : 1);

const transposePhraseChord = (
  chord: DefensePhraseChord,
  interval: string,
  minLowestNote: string,
): DefensePhraseChord => {
  const rawNames = chord.notes.map((note) => note.noteName);
  const transposedNames = interval === '1P'
    ? rawNames
    : rawNames.map((name) => transposeNoteName(name, interval));
  const repositioned = placeLowestInOctaveAbove(transposedNames, midiOf(minLowestNote));

  const notes = chord.notes.map((note, noteIndex) => {
    const noteName = repositioned[noteIndex] ?? note.noteName;
    const midi = midiOf(noteName);
    return {
      ...note,
      noteName,
      pitchMidi: midi,
      pitchClass: ((midi % 12) + 12) % 12,
      staff: staffForMidi(midi),
      staffChordName: note.staffChordName
        ? transposeChordSymbol(note.staffChordName, interval)
        : undefined,
    };
  });

  return {
    ...chord,
    chordName: transposeChordSymbol(chord.chordName, interval),
    notes,
  };
};

export const transposeDefensePhraseToKey = (
  template: DefensePhrase,
  referenceKey: MajorKey,
  targetKey: MajorKey,
  minLowestNote: string,
): DefensePhrase => {
  const interval = distance(referenceKey, targetKey);
  if (!interval) {
    throw new Error(`distance failed: ${referenceKey} -> ${targetKey}`);
  }
  const keyFifths = KEY_FIFTHS_BY_MAJOR[targetKey];
  const chords = template.chords.map((chord) => (
    transposePhraseChord(chord, interval, minLowestNote)
  ));
  return {
    ...template,
    title: targetKey,
    keyFifths,
    chords,
  };
};

export const buildTransposedVoicingPhrases = (
  stage: DefenseStage,
  keyState: DefenseVoicingKeyState,
): readonly DefensePhrase[] => {
  if (!isDefenseChordVoicingStage(stage)) {
    return stage.phrases;
  }
  const template = stage.phrases[0];
  if (!template) return stage.phrases;
  const targetKey = currentVoicingKey(keyState);
  const referenceKey = stage.voicingLowestKey ?? 'C';
  const minNote = stage.voicingMinLowestNote ?? 'C3';
  return [
    transposeDefensePhraseToKey(template, referenceKey, targetKey, minNote),
  ];
};
